### PR TITLE
Reuse preloaded lazy relations

### PR DESCRIPTION
**Pull request by StrikeForceZero**
Resolves https://github.com/typeorm/typeorm/issues/2965

### CODE REVIEW

File: test/github-issues/2965/entity/person.ts [hunk 1]
```diff
@@ -0,0 +1,14 @@
+import * as TypeOrm from '../../../../src/';
+import { Note } from './note';
+
+@TypeOrm.Entity()
```
[the remark this answers is not in the document]
Comment (StrikeForceZero) [reply test/github-issues/2965/entity/person.ts]: done in https://github.com/typeorm/typeorm/pull/2966/commits/0bd376be2de46e19906ff4f641567a79c59626ae

File: test/github-issues/2965/index.ts [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+
+        const originalLoad: (...args: any[]) => Promise<any[]> = connection.relationLoader.load;
+        let loadCalledCounter = 0;
+        connection.relationLoader.load = (...args: any[]): Promise<any[]> => {
+            loadCalledCounter++;
+            return originalLoad.call(connection.relationLoader, ...args);
+        }
+
+        {
+            console.log('preload notes');
```
Comment (Kononnable) [inline test/github-issues/2965/index.ts]: Let's clean these console logs - If test passes there is no need for it.

File: test/github-issues/2965/index.ts [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+            const res = await repoPerson.find({ relations: [ 'notes' ] });
+            console.log(res);
+            const personANotes = await res[ 0 ].notes;
```
Comment (Kononnable) [inline test/github-issues/2965/index.ts]: You use `[ 0 ]` here and `[0]` in line 50, it looks weird. Why different formatting? Same goes for lines 57 and 60.

**Comment by Kononnable** [comment]
Looks good for me except for what I pointed out earlier.

**Comment by pleerock** [comment]
Im not sure if we should continue to merge PRs related to "lazy relations feature" because I want to deprecate this experimental feature (since this feature brings a bad design and bring too much problems). What do you think @Kononnable

**Comment by Kononnable** [comment]
If work is already done we can merge it and eventually remove it in 0.3.

I think some people use this feature and will be disappointed if we remove it. There are very few situations where this feature really shines, but most of the times it is just to make developer happy in cost of performance(lazy loading in general).
Lazy loading could get rid of some of it problems if getters/setters could accept different types and convert them to property type. Last I checked such feature request is marked as open in typescript repo and there are no plans to implement it in the near future. Until then lazy loading isn't very useful for me, but some folks use it for some reasons.

**Comment by pleerock** [comment]
> I think some people use this feature and will be disappointed if we remove it.

of course they will be disappointed. But they will be disappointed even more if they choose the wrong way from the beginning and some day will simply stop using ORM because it will become too problematic for them

I can call current lazy loading implementation as an anti-pattern because it brings more problems then it solve.

**Comment by Kononnable** [comment]
I don't like how it is implemented too. The question is: will we be able to implement it differently(in eg. 1 year), or are we going to say it is to complex to be implemented correctly(language, orm limitations etc.) and this feature probably won't come back in future versions of typeorm.

**Comment by pleerock** [comment]
Its not about complexity, its about design and how good or bad this particular feature fits the design and philosophy of the framework. This particular feature can be deprecated because it imposes a bad design, it become to be a bad practice I would say.

The point of this feature is to load relation entities in some way, e.g. `const categories = await post.categories`. Same can be already done, maybe more verbose but basically same using relational query builder. We can simply deprecate this feature in favour of that syntax, and/or introduce a new additional syntax as I mentioned in #2902, something like `model.owner = repository.loadRelation("owner", model)`

File: test/github-issues/2965/index.ts (hunk 3 above)
Comment (StrikeForceZero) [reply test/github-issues/2965/index.ts]: copy/paste and intellij trying to be smart and format the pasted text. fixed.

File: test/github-issues/2965/index.ts (hunk 2 above)
Comment (StrikeForceZero) [reply test/github-issues/2965/index.ts]: all removed

**Comment by StrikeForceZero** [comment]
@pleerock I totally understand why lazy loading can create some massive headaches.

That being said, is there an active issue being used to discuss this?

I guess while I'm here, only thing I would really like to see, before it being removed, is a way to load a relationship for an entity similar to what you described:  `model.owner = repository.loadRelation("owner", model)`, maybe even something like:
```typescript
model.owner = repository.loadRelation("owner");
// these options would be great for type saftey and refactoring support
model.owner = modelRepository.loadRelation(model => model.owner);
model.owner = repository.loadRelation(model, model => model.owner);
// bonus points for Entity Manager
model.owner = entityManager.loadRelation(model, model => model.owner);
modelRelationLoader = entityManager.relationLoader(model);
model.owner = modelRelationLoader(model => model.owner);
// or.. maybe getting a little crazy
modelOwnerRelationLoader = entityManager.relationLoader(Model)(model => model.owner);
model.owner = modelOwnerRelationLoader(model);
```

**Comment by Kononnable** [comment]
@pleerock I didn't quite well put my thoughts into words. What I meant was that right now we have to remember about many things when using lazy loading. Some of these problems can be dealt with by using better design others will stay. We will never be able to achieve similar transparency as eg. Entity Framework because of language limitations. In EF there is no difference from user code perspective between using lazy loading and having each entity already loaded.

As for `model.owner = repository.loadRelation("owner", model)` - I don't like passing a string parameter, better to make use of type system. Also what if someone would try to use `model.owner` before it is loaded - should we return undefined or try to throw an error to make developer aware.

I don't think there is a issue with this proposal. Maybe we should create one and put ours ideas there - it will be easier to find and more people will be able to share their opinions.

**Comment by pleerock** [comment]
Okay, lets merge this one then. But @StrikeForceZero be aware, I think we'll have to deprecate lazy loading functionality in the future and replace with suggested or similar api.

**Pull request merged**